### PR TITLE
Ft search 159054021

### PR DESCRIPTION
**Pull request by Duncanian**
**What does this PR do?**

- Implements filter functionality by author, article title and tag

- Implements search functionality by keywords from author, article title, article body, article description and tag

**background Information.**
Enables a user to filter or search articles of their preference by author, article title, article body, article description and tag

**Install dependencies**
`pip install -r requirements.txt`

**How To Manually Test?**

Pull ft-search-159054021
`git pull ft-search-159054021`

Checkout to ft-search-159054021 branch
`git checkout ft-search-159054021`

Update database by running migrations.
   `python manage.py makemigrations`
   `python manage.py migrate`

Start the server
`python manage.py runserver`

Test the endpoints on postman.

Run test coverage
`coverage run manage.py test`
`coverage report`

**Affected Endpoints:**

`GET /api/articles`

***Filter***

**Filter by author**
`GET /api/articles?author__id=<author_id>`

**Filter by title**
`GET /api/articles?title=<title>`

**Filter by tag**
`GET /api/articles?tags__tag=<tag>`

RESPONSE
```
[
    {
        "title": "How to train your dragon 100",
        "slug": "how-to-train-your-dragon-100-5o62zifwg0",
        "body": "You have to believe",
        "description": "Ever wonder how?",
        "image_url": null,
        "created_at": "2018-08-15T10:56:55.808407Z",
        "updated_at": "2018-08-15T10:56:55.808433Z",
        "author": {
            "username": "ian2",
            "bio": "",
            "image": "https://static.productionready.io/images/smiley-cyrus.jpg"
        },
        "likes": [],
        "dislikes": [],
        "likes_count": 0,
        "dislikes_count": 0,
        "favorited": false,
        "favoriteCount": 0,
        "tagList": [
            "vikings",
            "dragon-training",
            "dragon",
            "dragons"
        ]
    }
]

```

***Search***

**Search by author, article title, article body, article description and tag **
`GET /api/articles?search=<keyword>`

RESPONSE
```
[
    {
        "title": "How to train your dragon 100",
        "slug": "how-to-train-your-dragon-100-5o62zifwg0",
        "body": "You have to believe",
        "description": "Ever wonder how?",
        "image_url": null,
        "created_at": "2018-08-15T10:56:55.808407Z",
        "updated_at": "2018-08-15T10:56:55.808433Z",
        "author": {
            "username": "ian2",
            "bio": "",
            "image": "https://static.productionready.io/images/smiley-cyrus.jpg"
        },
        "likes": [],
        "dislikes": [],
        "likes_count": 0,
        "dislikes_count": 0,
        "favorited": false,
        "favoriteCount": 0,
        "tagList": [
            "vikings",
            "dragon-training",
            "dragon",
            "dragons"
        ]
    }
]

```

What are the relevant pivotal tracker stories?

[#159054021](https://www.pivotaltracker.com/story/show/159054021) Add Search Functionality with custom filters

**Screenshots.**
screen shot 2018-08-15 at 18 10 24

screen shot 2018-08-15 at 19 42 01

image

image

### CODE REVIEW

**Comment by jacksono** [comment]
@Duncanian Merge the latest changes from develop into this branch. Also include a screen shot for a case when multiple query parameters are used eg searching by tags, author, article title ...etc

**Comment by dessHub** [comment]
@jacksono I have merged the branch with the develop and also added a screenshot.

**Comment by jacksono** [comment]
@dessHub the screenshot you have added is not quite what I meant. There should be multiple query parameter support. Something like .../article?author=jake&&title=geology` for cases when you are filtering by more than one field, or did you handle that in another way?

**Comment by dessHub** [comment]
@jacksono I get your point now. Uploading it now

**Review by dessHub — COMMENTED** [review]
.

**Comment by Duncanian** [comment]
@jacksono , am getting the screenshots for multiple query parameters.

**Comment by Duncanian** [comment]
@jacksono we have now handled the multiple query parameters.

**Comment by Tittoh** [comment]
Tested it and it works as expected

**Comment by jacksono** [comment]
@Duncanian @dessHub Its looking good now.  Apologies if i misled you but the multiple query params should be separated by `&` not `&&`. Fix that and update your branch. You also seem to have made changes to the `requirements.txt`. Please crosscheck and make sure only packages needed for this project are listed, and don't remove any packages that are needed.

**Comment by Duncanian** [comment]
@jacksono , I have made the requested changes.

File: requirements.txt [hunk 1]
```diff
@@ -1,13 +1,26 @@
...
+coverage==4.5.1
+coveralls==1.3.0
+defusedxml==0.5.0
 Django==2.0.7
 django-cors-middleware==1.3.1
 django-environ==0.4.5
 django-extensions==2.1.0
+django-filter==2.0.0
 djangorestframework==3.8.2
+docopt==0.6.2
```
Comment (asamisellah) [inline requirements.txt]: Is this package necessary? Are you using it anywhere?

**Review by Tittoh — DISMISSED** [review]

**Review by asamisellah — CHANGES_REQUESTED** [review]
When I run migrations there is an error because some modules are not installed. Kindly include instructions on your PR to pip install -r requirements.txt to prevent this.

File: requirements.txt (hunk 1 above)
Comment (Duncanian) [reply requirements.txt]: I didn't use it. Let me remove it too.

**Review by Duncanian — COMMENTED** [review]
@asamisellah ,I made the changes

**Review by asamisellah — APPROVED** [review]

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 """ Views for django Articles. """
 from django.shortcuts import render
 from django.db.models import Avg
-
+# from url_filter.integrations.drf import DjangoFilterBackend
```
Comment (denniswanjiru) [inline authors/apps/articles/views.py]: I think you should only push code that is not commented out according to the Andela Best Practices guideline

**Review by asamisellah — DISMISSED** [review]
I have tested the feature and it works as expected. It would be informative to have a 'Not found' message returned when an item is not found during a search. You don't have to but you may kindly consider.

**Comment by jacksono** [comment]
@Duncanian Look into these conflicts after merging in the pagination branch

**Comment by Duncanian** [comment]
@jacksono , the PR is ready for your final review.

**Comment by dessHub** [comment]
@jacksono All merged conflicts have been resolved.

**Review by jacksono — APPROVED** [review]

**Pull request merged**